extensions/eclipse-che-theia-terminal/src/node/workspace-service-impl.ts

```diff
@@ -26,80 +26,70 @@ export class CHEWorkspaceServiceImpl implements CHEWorkspaceService {
     constructor(@inject(EnvVariablesServer) protected readonly baseEnvVariablesServer: EnvVariablesServer) {
     }
 
-    public async getMachineList(): Promise<{ [attrName: string]: che.workspace.Machine }> {
-        const machineNames: { [attrName: string]: che.workspace.Machine } = {};
-        const workspaceId = await this.getWorkspaceId();
-        const restClient = await this.getRemoteApi();
-        if (!workspaceId || !restClient) {
-            return machineNames;
-        }
-        return new Promise<{ [attrName: string]: che.workspace.Machine }>((resolve, reject) => {
-            restClient.getById<che.workspace.Workspace>(workspaceId)
-                .then((workspace: che.workspace.Workspace) => {
-                    if (workspace.runtime) {
-                        resolve(workspace.runtime.machines);
-                        return;
+    public async getContainerList(): Promise<WorkspaceContainer[]> {
+        const containers: WorkspaceContainer[] = [];
+        try {
+            const workspaceId = await this.getWorkspaceId();
+            const restClient = await this.getRemoteApi();
+            if (!workspaceId || !restClient) {
+                throw new Error('Unable to use workspace client.');
+            }
+
+            const workspace = await restClient.getById<che.workspace.Workspace>(workspaceId);
```

Comment:
Only one try. Could you provide the possibility of processing request errors and add multiple attempts for getting data from API?

Reply:
In this case, container widget cannot open a new terminal from their own widget, because the same method is using for register command to creating a new terminal from the containers tree.

Reply:
For me multiple attempts looks ugly. If You really think that we should do it, we should do it it the one place - https://github.com/eclipse/che-workspace-client. And this hack became available for all plugins and extensions.

Reply:
You should think about it before approving this  PR
https://github.com/eclipse/che-theia/pull/13/files#diff-7906a66c8c2b4bf86faa2b57c70a2aaaR79
I don't see the problem to change this part. Is it principal for you? I can prepare changes for it WDYT?